python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,535 @@
+
+        xc_rotated, yc_rotated = self.rotate((xd_offset*(1+scale*r_distance_from_center/100),
+                                              yd_offset*(1+scale*r_distance_from_center/100)), (0, 0), rot)
+
+        xc_rotated = xc_rotated+x_scale_rot
+        yc_rotated = yc_rotated+y_scale_rot
+
+        xd_new = xc_rotated + xd_mod
+        yd_new = yc_rotated + yd_mod
+        return xd_new, yd_new
+
+    def quality_measure_single_spot(self, list_of_prepared_observations_and_predictions_single_fiber,
+                                    xd, yd):
+        """Calculate a penalty given x and y dot positions
+
+        Parameters
+        ----------
+        list_of_prepared_observations_and_predictions: `list`
+            list contaning observations for only one fiber
+        xd: `float`
+            x position of the dot
+        yd: `float`
+            y position of the dot
+
+        Returns
+        ----------
+        quality_measure: `float`
+            total penalty for this position of the dot
+
+        Notes
+        ---------
+        Calculates `penalty`. Penalty consist of 2 factors:
+            1. Points should have been seen given the dot position,
+            but they are not seen
+            2. Point which are seen, but they should not have been
+            seen, given the dot position
+
+        Calls `check`
+        Gets called by `total_penalty_for_single_dot`
+        """
+        list_of_quality_measure_single_run = []
+        for i in range(len(list_of_prepared_observations_and_predictions_single_fiber)):
+            prepared_observations_and_predictions =\
+                list_of_prepared_observations_and_predictions_single_fiber[i]
+
+            actual_position_x_observed,\
+                actual_position_y_observed,\
+                predicted_position_x_not_observed,\
+                predicted_position_y_not_observed = prepared_observations_and_predictions
+
+            # the ones which were seen - add penalty if you should not have seen them
+            quality_measure_observed = []
+            for j in range(len(actual_position_x_observed)):
+                quality_measure_observed.append(self.check(actual_position_x_observed[j],
+                                                           actual_position_y_observed[j], xd, yd, 0.75))
+
+            # the ones which should not be seen - add penalty if you saw them
+            quality_measure_predicted = []
+            for j in range(len(predicted_position_x_not_observed)):
+                # if you were not able to predict position, no penalty
+                if np.isnan(predicted_position_x_not_observed[j]):
+                    val = 0
+                else:
+                    val = self.check(predicted_position_x_not_observed[j],
+                                     predicted_position_y_not_observed[j], xd, yd, 0.75)
+                    val = 1 - val
+                quality_measure_predicted.append(val)
```

Comment:
You know how long this is going to be, so start with an array, and work with the entire array rather than individual elements.